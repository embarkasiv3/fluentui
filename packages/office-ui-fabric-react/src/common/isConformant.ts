import { isConformant as isConformantBase, IsConformantOptions } from '@fluentui/react-conformance';

export function isConformant(testInfo: IsConformantOptions) {
  const defaultOptions = {
    disabledTests: ['has-docblock'],
  };

  isConformantBase(defaultOptions, testInfo);
}
