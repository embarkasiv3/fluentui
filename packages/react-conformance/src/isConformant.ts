import { TestObject, TestingOptions } from './types';
import { withCustomConfig } from 'react-docgen-typescript';
import * as path from 'path';
import { defaultTests } from './defaultTests';

export function isConformant(testInfo: TestingOptions, extraTests?: TestObject) {
  const { componentPath, displayName } = testInfo;
  const tsconfigPath = path.join(process.cwd(), 'tsconfig.json');
  const parser = withCustomConfig(tsconfigPath, {});
  const components = parser.parse(componentPath);
  const mainComponents = components.filter(comp => comp.displayName === displayName);

  if (mainComponents.length === 1) {
    const componentInfo = mainComponents[0];

    for (const test of Object.keys(defaultTests)) {
      defaultTests[test](componentInfo, testInfo);
    }
    if (extraTests) {
      for (const test of Object.keys(extraTests)) {
        extraTests[test](componentInfo, testInfo);
      }
    }
  } else {
    if (components.length === 0) {
      throw new Error('No exported components in path: ' + componentPath);
    } else {
      throw new Error(
        `No component with name '${displayName}' was found at ${componentPath}. ` +
          `These are the exported component names: ${components.map(component => component.displayName).join(', ')}`,
      );
    }
  }
}
