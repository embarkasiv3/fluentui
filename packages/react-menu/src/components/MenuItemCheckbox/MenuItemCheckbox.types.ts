import * as React from 'react';
import { ComponentProps, ObjectShorthandProps, ShorthandProps } from '@fluentui/react-utils';
import { MenuItemSelectableProps, MenuItemSelectableState } from '../../selectable/index';
import { MenuItemProps, MenuItemState } from '../MenuItem/MenuItem.types';

/**
 * {@docCategory MenuItemCheckbox}
 */
export interface MenuItemCheckboxProps
  extends ComponentProps,
    React.HTMLAttributes<HTMLElement>,
    MenuItemProps,
    MenuItemSelectableProps {
  /**
   * Icon slot rendered before children content
   */
  icon?: ShorthandProps<HTMLElement>;

  /**
   * Slot for the checkmark indicator
   */
  checkmark?: ShorthandProps<HTMLElement>;
}

/**
 * {@docCategory MenuItemCheckbox}
 */
export interface MenuItemCheckboxState extends MenuItemCheckboxProps, MenuItemState, MenuItemSelectableState {
  ref: React.MutableRefObject<HTMLElement>;

  /**
   * Icon slot rendered before children content
   */
  icon?: ObjectShorthandProps<HTMLElement>;

  /**
   * Slot for the checkmark indicator
   */
  checkmark: ObjectShorthandProps<HTMLElement>;
}
