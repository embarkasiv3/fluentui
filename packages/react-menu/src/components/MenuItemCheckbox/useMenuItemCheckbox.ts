import * as React from 'react';
import { makeMergeProps, resolveShorthandProps } from '@fluentui/react-utilities';
import { MenuItemCheckboxProps, MenuItemCheckboxState } from './MenuItemCheckbox.types';
import { useMenuItemSelectable } from '../../selectable/index';
import { useMergedRefs } from '@fluentui/react-hooks';

/**
 * Consts listing which props are shorthand props.
 */
export const menuItemCheckboxShorthandProps = ['icon', 'checkmark'];

const mergeProps = makeMergeProps<MenuItemCheckboxState>({ deepMerge: menuItemCheckboxShorthandProps });

/** Returns the props and state required to render the component */
export const useMenuItemCheckbox = (
  props: MenuItemCheckboxProps,
  ref: React.Ref<HTMLElement>,
  defaultProps?: MenuItemCheckboxProps,
): MenuItemCheckboxState => {
  const state = mergeProps(
    {
      ref: useMergedRefs(ref, React.useRef(null)),
      icon: { as: 'span' },
      checkmark: { as: 'span' },
      role: 'menuitemcheckbox',
      tabIndex: 0,
    },
    defaultProps,
    resolveShorthandProps(props, menuItemCheckboxShorthandProps),
  );

  useMenuItemSelectable(state, () => {
    const newCheckedItems = [...state.checkedItems];
    const index = state.checkedItems.indexOf(state.value);
    if (index !== -1) {
      newCheckedItems.splice(index, 1);
    } else {
      newCheckedItems.push(state.value);
    }

    return newCheckedItems;
  });

  return state;
};
