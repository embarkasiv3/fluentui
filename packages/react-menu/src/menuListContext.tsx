import * as React from 'react';
import { createContext, useContextSelector, ContextSelector } from '@fluentui/react-context-selector';
import { SelectableHandler } from './selectable/index';

const MenuListContext = createContext<MenuListContextValue>({
  checkedValues: {},
  onCheckedValueChange: () => null,
  toggleCheckbox: () => null,
  selectRadio: () => null,
});

/**
 * Context shared between MenuList and its children components
 */
export interface MenuListContextValue {
  checkedValues?: Record<string, string[]>;
  onCheckedValueChange?: (e: React.MouseEvent | React.KeyboardEvent, name: string, items: string[]) => void;
  toggleCheckbox?: SelectableHandler;
  selectRadio?: SelectableHandler;
}

export const MenuListProvider = MenuListContext.Provider;

export const useMenuListContext = <T,>(selector: ContextSelector<MenuListContextValue, T>) =>
  useContextSelector(MenuListContext, selector);
