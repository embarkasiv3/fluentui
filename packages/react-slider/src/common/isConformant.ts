import { isConformant as baseIsConformant, IsConformantOptions } from '@fluentui/react-conformance';

export function isConformant(testInfo: IsConformantOptions) {
  const defaultOptions = {
    disabledTests: ['has-docblock', 'kebab-aria-attributes'],
  };

  baseIsConformant(defaultOptions, testInfo);
}
