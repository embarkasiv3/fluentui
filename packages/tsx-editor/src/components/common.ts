/** Font family for code snippets/editors */
export const codeFontFamily = 'Monaco, Menlo, Consolas, "Droid Sans Mono", "Inconsolata", "Courier New", monospace';
