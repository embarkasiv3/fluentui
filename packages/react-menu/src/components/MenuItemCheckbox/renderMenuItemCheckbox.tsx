import * as React from 'react';
import { getSlots } from '@fluentui/react-utilities';
import { MenuItemCheckboxState } from './MenuItemCheckbox.types';
import { menuItemCheckboxShorthandProps } from './useMenuItemCheckbox';

/** Function that renders the final JSX of the component  */
export const renderMenuItemCheckbox = (state: MenuItemCheckboxState) => {
  const { slots, slotProps } = getSlots(state, menuItemCheckboxShorthandProps);

  return (
    <slots.root {...slotProps.root}>
      <slots.checkmark {...slotProps.checkmark} />
      <slots.icon {...slotProps.icon} />
      {state.children}
    </slots.root>
  );
};
