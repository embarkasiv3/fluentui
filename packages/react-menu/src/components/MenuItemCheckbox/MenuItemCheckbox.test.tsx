import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { render, fireEvent } from '@testing-library/react';
import { ReactWrapper } from 'enzyme';
import { isConformant } from '../../common/isConformant';
import { MenuItemCheckbox } from './MenuItemCheckbox';
import { MenuListContext, MenuListProvider } from '../../menuListContext';

describe('MenuItemCheckbox conformance', () => {
  isConformant({
    asPropHandlesRef: true,
    // TODO fix generics in conformance
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    Component: MenuItemCheckbox,
    requiredProps: {
      name: 'checkbox',
      value: '1',
    },
    displayName: 'MenuItemCheckbox',
  });

  let wrapper: ReactWrapper | undefined;

  afterEach(() => {
    if (wrapper) {
      wrapper.unmount();
      wrapper = undefined;
    }
  });

  /**
   * Note: see more visual regression tests for MenuItemCheckbox in /apps/vr-tests.
   */
  it('renders a default state', () => {
    const component = renderer.create(
      <MenuItemCheckbox name="checkbox" value="1">
        Default MenuItemCheckbox
      </MenuItemCheckbox>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('MenuItemCheckbox', () => {
  const TestMenuListContext = (props: { children: React.ReactNode; context?: Partial<MenuListContext> }) => {
    const contextValue: MenuListContext = {
      checkedValues: {},
      onCheckedValueChange: jest.fn(),
      ...(props.context && props.context),
    };

    return <MenuListProvider value={contextValue}>{props.children}</MenuListProvider>;
  };

  it('should render checkmark slot if checked', () => {
    // Arrange
    const checkedValues = { test: ['1'] };
    const checkmark = 'xxx';
    const { getByText } = render(
      <TestMenuListContext context={{ checkedValues }}>
        <MenuItemCheckbox name="test" value="1" checkmark={checkmark}>
          Checkbox
        </MenuItemCheckbox>
      </TestMenuListContext>,
    );

    // Assert
    expect(getByText(checkmark)).not.toBeNull();
  });

  it('should render icon slot', () => {
    // Arrange
    const icon = 'xxx';
    const { getByText } = render(
      <TestMenuListContext>
        <MenuItemCheckbox name="test" value="1" icon={icon}>
          Checkbox
        </MenuItemCheckbox>
      </TestMenuListContext>,
    );

    // Assert
    expect(getByText(icon)).not.toBeNull();
  });

  it('should set aria-checked value to true if value is checked', () => {
    // Arrange
    const checkedValues = { test: ['1'] };
    const checkmark = 'xxx';
    const { container } = render(
      <TestMenuListContext context={{ checkedValues }}>
        <MenuItemCheckbox name="test" value="1" checkmark={checkmark}>
          Checkbox
        </MenuItemCheckbox>
      </TestMenuListContext>,
    );

    // Assert
    expect(container.querySelector('[role="menuitemcheckbox"]')?.getAttribute('aria-checked')).toEqual('true');
  });

  it.each([
    ['uncheck', ['1'], []],
    ['check', [], ['1']],
  ])('should %s checkbox on click', (_, checkedItems, expectedResult) => {
    // Arrange
    const checkboxName = 'name';
    const checkedValues = { [checkboxName]: checkedItems };
    const spy = jest.fn();
    const { container } = render(
      <TestMenuListContext context={{ checkedValues, onCheckedValueChange: spy }}>
        <MenuItemCheckbox name={checkboxName} value={'1'}>
          Checkbox
        </MenuItemCheckbox>
      </TestMenuListContext>,
    );

    // Act
    const menuitem = container.querySelector('[role="menuitemcheckbox"]');
    menuitem && fireEvent.click(menuitem);

    // Assert
    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy).toHaveBeenCalledWith(expect.anything(), checkboxName, [...expectedResult]);
  });
});
