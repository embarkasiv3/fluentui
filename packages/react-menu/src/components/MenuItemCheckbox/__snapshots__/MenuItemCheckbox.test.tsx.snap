// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuItemCheckbox conformance renders a default state 1`] = `
<div
  aria-checked={false}
  className=""
  name="checkbox"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="menuitemcheckbox"
  tabIndex={0}
>
  Default MenuItemCheckbox
</div>
`;
